### Overview
NewbornRisk location on server: `/mnt/storage_pool/Genomics/Genome/NewbornRisk`
1. Script for batch samples analysis: `/mnt/storage_pool/Genomics/Genome/NewbornRisk/NewbornRisk_batch.sh`
2. Example output files: `/mnt/storage_pool/Genomics/Genome/NewbornRisk/examples/newborn103`
3. Input file for example:
  - Merged VCF file: `/mnt/storage_pool/Genomics/Genome/NewbornRisk/examples/newborn103/newborn103_merged.vcf.gz`
  - SampleIDs and Genders: `/mnt/storage_pool/Genomics/Genome/NewbornRisk/examples/newborn103/newborn103gender.txt`

### Command for anlaysis
```bash
newbornrisk='/mnt/storage_pool/Genomics/Genome/NewbornRisk'
bash NewbornRisk_batch.sh \
  -i newborn103 \
  -v $newbornrisk/examples/P0064/newborn103_merged.vcf.gz \
  -o $newbornrisk/examples/P0064 \
  --sample-metadata $newbornrisk/examples/P0064/newborn103gender.txt \
  --genome GRCH38 --only-pass yes --fork 48 --threads 20 --genedb TR
```

## UI Design

### Summary 
- Overview - Cohort summary: `Cohort_summary.txt`
- Overview - Screened Variant Statistic: `monogenic_carrier_count.txt` and Figure `Figures/monogenic_carrier_count.pdf`
- Screening Result Summary:
  - Positive Monogenic Diseases: `monogenic_positive_results.txt`
  - Carrier Status: `carrier_status_results.txt`
  - Figure for positive monogenic diseases: `Figures/Monogenic_details.png`
- Disease Ontology:
  - Positive Monogenic Diseases: `Positive.ontology.count.txt`
  - Carrier Status: `Carrrier.ontology.count.txt`
  - Ontology Radar plot: `Figures/Positive_Carrier_Ontology.png`
- Cross-Study Comparisons (Venn plots):
  - Positive Monogenic Diseases: `Figures/Positive_comparison_venn.png`
  - Carrier Status: `Figures/Carrier_comparison_venn.png`
- Detailed Comparisons:
  - Genes of positive monogenic diseases: `Positive_comparison_venn_table.txt` 
  - Genes of carrier status: `Carrier_comparison_venn_table.txt`
  - Positive monogenic diseases shared with BabySeq, EarlyCheck, and BabyDetect: `compare_positive_results.txt`
  - Positive monogenic diseases shared with Guardian: `compare_guardian_positive_results.txt`
  - Carrier status shared with BabySeq: `compare_babyseq_carrier_results.txt`
  - Carrier status shared with EarlyChec: `compare_earlycheck_carrier_results.txt`


<img width="887" height="662" alt="image" src="https://github.com/user-attachments/assets/2c52380a-966a-4a4a-85bc-02607de819f5" />
<img width="877" height="602" alt="image" src="https://github.com/user-attachments/assets/58499410-8405-4685-876b-8b3198f31d80" />
<img width="1670" height="493" alt="image" src="https://github.com/user-attachments/assets/8356c5dc-4076-4bdb-bc73-b4a254f78cea" />

### Variant Statistics
Tables and figures are generated from VEP: `newborn103_vep_annotated.vcf.gz_summary.html` \
We can directly use these resutls and show in our web server. 

### Vairnat Details
- Variant Detail table: `newborn103_vep_merged_rmmissingalt_biallelic2.txt`
<img width="1671" height="908" alt="image" src="https://github.com/user-attachments/assets/a88217e8-b8e3-48b3-97ec-0425e16a2371" />

### PGx 
- Haplotypes statistics: `PGx_statistics.txt`
- PGx haplotype distribution: `PGx_samples_variants_number.txt` and Figure `Figures/PGx_count.png`
- Individual PGx Results: every single files in `PGx/`
<img width="876" height="518" alt="image" src="https://github.com/user-attachments/assets/e871abfc-f74d-4ae6-82e3-c609446cdab9" />


